editor/src/components/canvas/canvas-strategies/strategies/paste-metastrategy.tsx

```diff
@@ -0,0 +1,254 @@
+import {
+  StaticReparentTarget,
+  reparentStrategyForPaste,
+} from './reparent-helpers/reparent-strategy-helpers'
+import { elementToReparent } from './reparent-utils'
+import { updateFunctionCommand } from '../../commands/update-function-command'
+import { foldAndApplyCommandsInner } from '../../commands/commands'
+import { updateSelectedViews } from '../../commands/update-selected-views-command'
+import { MetadataUtils } from '../../../../core/model/element-metadata-utils'
+
+const PasteModes = ['replace', 'preserve'] as const
+type PasteMode = typeof PasteModes[number]
+
+export const PasteWithPropertiesReplacedStrategyId = 'PasteWithPropertiesReplacedStrategy'
+export const PasteWithPropertiesPreservedStrategyId = 'PasteWithPropertiesPreservedStrategy'
+
+export const pasteWithPropsReplacedStrategy = pasteMetaStrategy('replace')
+export const pasteWithPropsPreservedStrategy = pasteMetaStrategy('preserve')
+
+type StrategySlice = Pick<CanvasStrategy, 'fitness' | 'id' | 'name'>
+
+export function pasteStrategy(
+  interactionSession: InteractionSession,
+  canvasState: InteractionCanvasState,
+  elementPasteWithMetadata: ElementPasteWithMetadata,
+  strategyProps: StrategySlice,
+): CanvasStrategy {
+  return {
+    id: strategyProps.id,
+    name: strategyProps.name,
+    fitness: strategyProps.fitness,
+    controlsToRender: [],
+    apply: () => {
+      if (
+        interactionSession?.interactionData.type !== 'STATIC_REPARENT' ||
+        canvasState.interactionTarget.type !== 'TARGET_PATHS'
+      ) {
+        return emptyStrategyApplicationResult
+      }
+
+      const target = getTargetParentForPaste(
+        canvasState.projectContents,
+        canvasState.interactionTarget.elements,
+        canvasState.nodeModules,
+        canvasState.openFile,
+        canvasState.startingMetadata,
+        interactionSession.interactionData.pasteTargetsToIgnore,
+        {
+          elementPaste: elementPasteWithMetadata.elements,
+          originalContextMetadata: elementPasteWithMetadata.targetOriginalContextMetadata,
+          originalContextElementPathTrees:
+            interactionSession.interactionData.targetOriginalPathTrees,
+        },
+        canvasState.startingElementPathTree,
+      )
+      if (target == null) {
+        return emptyStrategyApplicationResult
+      }
+
+      const elements = getElementsFromPaste(
+        elementPasteWithMetadata.elements,
+        target.parentPath,
+        canvasState.projectContents,
+      )
+
+      const strategy = reparentStrategyForPaste(
+        canvasState.startingMetadata,
+        canvasState.startingAllElementProps,
+        canvasState.startingElementPathTree,
+        target.parentPath.intendedParentPath,
+      )
+
+      const commands = elements.flatMap((currentValue) => {
+        return [
+          updateFunctionCommand('always', (editor, commandLifecycle) => {
+            if (interactionSession.interactionData.type !== 'STATIC_REPARENT') {
```

Comment:
nit: if you put interactionSession.interactionData in a const and then type check that, TS won't bother you with later property accesses.